prow/cmd/deck/main.go

```diff
@@ -408,6 +421,144 @@ func handleBadge(ja *jobs.JobAgent) http.HandlerFunc {
+			}
+
+		} else {
+			src = r.URL.Query().Get("src")
+			if src == "" {
+				http.Error(w, "missing src query parameter", http.StatusBadRequest)
+				return
+			}
+		}
+
+		artifactNames, err := sg.ListArtifacts(src)
+		if err != nil {
+			logrus.WithError(err).Error("Error listing artifacts")
+		}
+
+		viewerCache := map[string][]string{}
+
+		for re, viewerNames := range ca.Config().Deck.Spyglass.Viewers {
+			matches := []string{}
+			r, err := regexp.Compile(re)
+			if err != nil {
+				logrus.WithError(err).Error("Regexp failed to compile.")
+				continue
+			}
+			for _, a := range artifactNames {
+				if r.MatchString(a) {
+					matches = append(matches, a)
+				}
+			}
+			for _, vName := range viewerNames {
+				viewerCache[vName] = matches
+			}
+		}
+
+		viewsStart := time.Now()
+		lenses := sg.Views(viewerCache)
+		viewsElapsed := time.Since(viewsStart)
+		logrus.Info("Got views in ", viewsElapsed)
+
+		var viewBuf bytes.Buffer
+		type ViewsTemplate struct {
+			Views       []spyglass.Lens
+			Source      string
+			ViewerCache map[string][]string
+		}
+		vTmpl := ViewsTemplate{
+			Views:       lenses,
+			Source:      src,
+			ViewerCache: viewerCache,
+		}
+		t := template.Must(template.ParseFiles("/static/template/spyglass-template.html"))
+		tErr := t.Execute(&viewBuf, vTmpl)
+		if tErr != nil {
+			logrus.WithError(tErr).Error("Error rendering template.")
+		}
+
+		fmt.Fprint(w, viewBuf.String())
+		elapsed := time.Since(start)
+		logrus.Info("Time taken to return view page: ", elapsed)
+	}
+}
+
+// handleArtifactView handles requests to load a single view for a job
+// Query params:
+// - name: required, specifies the name of the viewer to load
+// - src: required, specifies the job source from which to fetch artifacts
+func handleArtifactView(sg *spyglass.Spyglass) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		start := time.Now()
+		setHeadersNoCaching(w)
+		w.Header().Set("Content-Type", "application/json")
+		name := r.URL.Query().Get("name")
+		src := r.URL.Query().Get("src")
+		if name == "" {
+			http.Error(w, "missing name query parameter", http.StatusBadRequest)
+			return
+		}
+		if src == "" {
+			http.Error(w, "missing src query parameter", http.StatusBadRequest)
+			return
+		}
+
+		body, err := ioutil.ReadAll(r.Body)
+		if err != nil {
+			http.Error(w, "failed to read body", http.StatusBadRequest)
+			return
+		}
+
+		viewReq := &spyglass.ViewRequest{}
+		err = json.Unmarshal(body, viewReq)
+		if err != nil {
+			http.Error(w, "failed to marshal request body", http.StatusBadRequest)
+		}
+
+		lens := sg.Refresh(src, "", viewReq)
+		pd, err := json.Marshal(lens)
+		if err != nil {
+			logrus.WithError(err).Error("Error marshaling payload.")
+			pd = []byte("{}")
+		}
+		fmt.Fprint(w, string(pd))
+		elapsed := time.Since(start)
+		logrus.Infof("Time taken to refresh %s: %s", name, elapsed) //TODO (paulangton): move these load times next to the title of the viewer
```

Comment:
more comprehensive TODO is to expose as a prometheus metric, not necessarily in the log

Reply:
fixed in latest